### DISCUSSION TITLE
how to use a pretrained model performance and also training the pretrained model with new dataset

### DISCUSSION THREAD

**Discussion by Premtheja31** [mindee/doctr#1311] [Q&A]
hi
how to use a pretrained model performance along with adding new symbols to vocabulary and train the symbols to already trained recognition model if we are doing with the below command 
python references/recognition/train_pytorch.py crnn_vgg16_bn --train_path train_path_rupee --val_path val_path --epochs 3 --device 1 --pretrained
it is performing worse than the pretrained model

if we use this below command 
 python references/recognition/train_pytorch.py crnn_vgg16_bn --train_path train_path_rupee --val_path val_path --epochs 3 --device 1 --resume crnn_vgg16_bn_original.pt
 it is showing error of weights and bias by the way crnn_vgg16_bn_original.pt this is the cache checkpoint of pretrained model 
 
 how to solve this error or else how to do the process ?

**Comment by Premtheja31** [comment]
![Screenshot (209)](https://github.com/mindee/doctr/assets/115713462/3738527e-6703-49f4-882f-42ea0d380aff)

**Reply by xReniar** [reply]
Is this the output for this command?
`python references/recognition/train_pytorch.py crnn_vgg16_bn --train_path train_path_rupee --val_path val_path --epochs 3 --device 1 --resume crnn_vgg16_bn_original.pt`

It seems that you added new characters on the vocabolary, so it gives you error because the pretrained models does not recognize the new characters. I had the same problem a few days ago, so I just started training with a new model without resuming with the pretrained model

**Reply by felixT2K** [reply]
@Premtheja31 
The first command is correct for finetuning the model.
Currently we reset only the last linear layer (classifier) but let the whole model trainable if it's trained on a other vocab as french.
What you can do is modifying the training script slightly by freezing all layers up to the last layer.
In this case you would only train the classifier with your new vocab.
NOTE: This will only work for all crnn models and vitstr models because the other models (parseq, sar, master) contain embedding layer

**Reply by felixT2K** [reply]
Snippet:

```from doctr.models import crnn_vgg16_bn

model = crnn_vgg16_bn(pretrained=True)

# Freeze all layers except the last one
for name, param in model.named_parameters():
    if name in ["linear.weight", "linear.bias"]:
        param.requires_grad = True
        print(f"Set layer {name} to trainable")
        continue
    param.requires_grad = False


for name, param in model.named_parameters():
    print(name, param.requires_grad)
```

https://github.com/mindee/doctr/blob/7ab0ece3a4aec256bb36c311f4c6eeaa597d826b/references/recognition/train_pytorch.py#L235

```
model = recognition.__dict__[args.arch](pretrained=args.pretrained, vocab=vocab)
# Freeze all layers except the last one
for name, param in model.named_parameters():
    if name in ["linear.weight", "linear.bias"]:  # specific layer for crnn_vgg16_bn
        param.requires_grad = True
        print(f"Set layer {name} to trainable")
        continue
    param.requires_grad = False
```
